### PR TITLE
fix: sharing query issue in JdbcEventStore

### PR DESCRIPTION
**Pull request by vietnguyen**

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/JpaQueryUtils.java [hunk 1]
```diff
@@ -341,8 +344,32 @@ public static String createSelectOrderExpression( @Nullable List<org.hisp.dhis.q
         };
     }
 
+    /**
+     *  Return SQL query for checking sharing access for given user
+     * @param sharingColumn sharing column reference
+     * @param user User for sharing checking
+     * @return SQL query
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/JpaQueryUtils.java:351]: Missing description of "access"

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/JpaQueryUtils.java [hunk 2]
```diff
@@ -341,8 +344,32 @@ public static String createSelectOrderExpression( @Nullable List<org.hisp.dhis.q
...
+    public static final String generateSQlQueryForSharingCheck( String sharingColumn, User user, String access )
+    {
+        List<Long> userGroupIds = getIdentifiers( user.getGroups() );
+
+        String groupsIds = CollectionUtils.isEmpty( userGroupIds ) ? null
+            : "{" + org.apache.commons.lang3.StringUtils.join( user.getGroups().stream().map( g -> g.getUid() )
+                .collect( Collectors.toList() ), "," ) + "}";
+
+        return  " ( " + sharingColumn + "->>'owner' is not null and " + sharingColumn + "->>'owner' = '" + user.getUid() + "') "
+            + " or " + sharingColumn + "->>'public' like '__r%' or " + sharingColumn + "->>'public' is null "
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/JpaQueryUtils.java:362]: If access is the access string, does it make sense to replace __r% with acccess here?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/JpaQueryUtils.java [hunk 3]
```diff
@@ -341,8 +344,32 @@ public static String createSelectOrderExpression( @Nullable List<org.hisp.dhis.q
...
+     */
+    public static final String generateSQlQueryForSharingCheck( String sharingColumn, User user, String access )
+    {
+        List<Long> userGroupIds = getIdentifiers( user.getGroups() );
+
+        String groupsIds = CollectionUtils.isEmpty( userGroupIds ) ? null
+            : "{" + org.apache.commons.lang3.StringUtils.join( user.getGroups().stream().map( g -> g.getUid() )
+                .collect( Collectors.toList() ), "," ) + "}";
+
+        return  " ( " + sharingColumn + "->>'owner' is not null and " + sharingColumn + "->>'owner' = '" + user.getUid() + "') "
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/JpaQueryUtils.java:361]: Would it be possible to rewrite this using String.format, or similar to make it a little bit more readable?

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/SpringDataValueSetStore.java [hunk 4]
```diff
@@ -374,10 +377,13 @@ private String getAttributeOptionComboClause( User user )
             .collect( Collectors.toList() );
 
         String groupAccessCheck =  groupsIds.size() > 0 ?
-            "or ( " +
-                "check_user_group_ids( co.sharing, '{"+ String.join( ",", groupsIds ) +"}' ) = true " +
-                "and check_user_groups_access( co.sharing, '__r%', '{"+ String.join( ",", groupsIds ) +"}' ) = true ) )"
-            : ") )";
+            " or ( " + JsonbFunctions.HAS_USER_GROUP_IDS + "( co.sharing, '{"+ String.join( ",", groupsIds ) + "}' ) = true " +
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/SpringDataValueSetStore.java:380]: Consider using String.format or similar here as well just to make it a bit more readable.

**Review by stian-sandvold — CHANGES_REQUESTED** [review]

**Review by mortenoh — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/SpringDataValueSetStore.java (hunk 4 above)
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/datavalueset/SpringDataValueSetStore.java]: fixed

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/JpaQueryUtils.java (hunk 3 above)
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/JpaQueryUtils.java]: fixed

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/JpaQueryUtils.java (hunk 2 above)
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/JpaQueryUtils.java]: fixed

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/JpaQueryUtils.java (hunk 1 above)
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/query/JpaQueryUtils.java:370]: fixed

**Pull request merged**